File: python/i2c.py
#!/usr/bin/env python3.8
import os
from time import sleep
# this stops windows interpreter from complaining
if os.name == "posix":
    from fcntl import ioctl

# TODO
# - implement external config file from json or cmdline params
# - document all functions


# Configure static parameters here for now.

i2c_bus = 1         # i2c bus to use. default to /dev/i2c-1
sensor_addr = 0x5c  # i2c address of the device
func_code = 0x03    # always 0x03 on AM2320.
reg_len = 0x04      # AM2320 has total 4 registers for humi and temp data.
reg_start = 0x00    # start reading from zero by default


class I2C(object):
    def __init__(self, bus: int = i2c_bus):
        self.dev = f"/dev/i2c-{bus}"
        self.fd = os.open(self.dev, os.O_RDWR)
        self.write_cmd = bytes([func_code, reg_start, reg_len])

    def __enter__(self):
        return self

    def __exit__(self):
        if self.fd:
            os.close(self.fd)

    def close(self):
        os.close(self.fd)

    def getData(self, addr: int = sensor_addr):
        if not self.fd:
            raise IOError(f"{self.dev} is not open")
        ioctl(self.fd, 0x0703, addr)

        try:
            os.write(self.fd, bytes(0x00))
        except Exception:
            pass
        sleep(0.0010)

        os.write(self.fd, self.write_cmd)
        sleep(0.0017)

        data = os.read(self.fd, 8)
        self.close()

        valid_data = self.__validate(data)

        humi = self.__merge(valid_data[2], valid_data[3])
        temp = self.__merge(valid_data[4], valid_data[5])

        if (temp & 0x8000):
            temp = -(temp & 0x7fff)

        return dict({
            f"humidity": humi/10,
            f"temperature": temp/10
        })

    def __validate(self, data: bytes):
        """
        Document this!
        """

        if data[0] != self.write_cmd[0] and data[1] != self.write_cmd[2]:
            raise Exception("ACK mismatch")
        crc = self.__merge(data[7], data[6])
        if self.__crc16(data[0:6]) != crc:
            raise Exception("CRC mismatch")
        return data

    @staticmethod
    def __crc16(data: bytes):
        crc = 0xFFFF

        for i in data:
            crc = crc ^ i
            for bit in range(8):
                if (crc & 0x0001):
                    crc >>= 1
                    crc ^= 0xA001
                else:
                    crc >>= 1
        return crc

    @staticmethod
    def __merge(a: int, b: int, numbits: int = 8):
        return (a << numbits) | b

File: c/i2c.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <linux/i2c-dev.h>
#include <time.h>

unsigned short crc16(unsigned char *ptr, unsigned char len)
{
    unsigned short crc = 0xFFFF;
    unsigned char i;
    while (len--)
    {
        crc ^= *ptr++;
        for (i = 0; i < 8; i++)
        {
            if (crc & 0x01)
            {
                crc >>= 1;
                crc ^= 0xA001;
            }
            else
            {
                crc >>= 1;
            }
        }
    }
    return crc;
}

int main()
{
    time_t timer;
    char t_buffer[20];
    struct tm *tm_info;

    int file_i2c;
    int len;
    unsigned char buf[8];

    // open i2c bus in RW mode
    //////////////////////////////////////
    char *fname = (char *)"/dev/i2c-1";

    if ((file_i2c = open(fname, O_RDWR)) < 0)
    {
        printf("failed to open i2c");
        exit(1);
    }

    // open i2c device
    //////////////////////////////////////
    int addr = 0x5c;
    if (ioctl(file_i2c, I2C_SLAVE, addr) < 0)
    {
        printf("failed to communicate with i2c device");
        exit(1);
    }
    usleep(850);

    // wake up sensor by sending something
    //////////////////////////////////////
    len = 1;
    buf[0] = 0x00;
    write(file_i2c, buf, len);

    usleep(850);

    // ask for data
    //////////////////////////////////////
    buf[0] = 0x03; // opening command
    buf[1] = 0x00; // start register
    buf[2] = 0x04; // num registers
    len = 3;
    if (write(file_i2c, buf, len) != len)
    {
        printf("failed to write");
        exit(1);
    }

    usleep(1700);

    // read data and write to file
    //////////////////////////////////////
    FILE *f = fopen("data.txt", "w");
    fflush(f);
    fclose(f);

    FILE *file = fopen("data.txt", "a");
    if (file == NULL)
    {
        printf("Error opening file\n");
        exit(1);
    }

    len = 8;
    if (read(file_i2c, buf, len) != len)
    {
        printf("failed to read i2c bus");
        exit(1);
    }
    else
    {
        int i, temp;
        unsigned short humi, checksum, crc;

        humi = (buf[2] << 8) | buf[3];
        temp = (buf[4] << 8) | buf[5];
        crc = (buf[7] << 8) | buf[6]; // am2320 sends crc "backwards"

        // negative temp is determined if bit 15 is 1
        if (temp & 0x8000)
        {
            temp = -(temp & 0x7fff);
        }

        checksum = crc16(buf, 6);

        if (checksum == crc)
        {
            timer = time(NULL);
            tm_info = localtime(&timer);

            strftime(t_buffer, 20, "%Y-%m-%d %H:%M:%S", tm_info);

            printf("RH: %.1f\nTemp: %.1f\nCRC: %02x\n", (float)humi / 10.0f, (float)temp / 10.0f, crc);
            printf("Timestamp: %s\n\n", t_buffer);

            fprintf(file, "\"%.1f\",\"%.1f\",\"%s\"", (float)humi / 10.0f, temp / 10.0f, t_buffer);

            /*
            for (i = 0; i < len; i++)
            {
                printf("%02x ", buf[i]);
            }
            */
            printf("\n");
        }
        else
        {
            printf("Checksum error.\nsensor: %04x\ncalculated: %04x\n\ndata: ", crc, checksum);

            for (i = 0; i < len; i++)
            {
                printf("%02x ", buf[i]);
            }
            printf("\n");

            exit(1);
        }
    }

    fclose(file);
    sleep(0.05);

    return 0;
}
